File: packages/react-native-ekyc/src/liveness/LivenessSession.ts
import {
  DEFAULT_SESSION_OPTIONS,
  type FaceSignal,
  type FailureReason,
  type Framing,
  type LivenessState,
  type SessionEvent,
  type SessionOptions,
  type StepMetric,
} from '../types'
import type { Challenge } from './Challenge'

/** Largest time step we trust between two frames; guards against a stalled camera. */
const MAX_FRAME_DELTA_MS = 250

/**
 * Drives the user through the challenge list and decides when to take a photo.
 *
 * Pure TypeScript: no React, no react-native, no timers, no `Date.now()`. All
 * time comes from `signal.t`, so a test can replay a whole session in a loop
 * and assert on every transition.
 *
 * Each step is a *hold*: the challenge predicate must stay true for
 * `holdMs`. Halfway through the hold the session emits a `capture` event —
 * the pose is certainly being held at that moment, which is what makes a still
 * photo valid evidence despite shutter latency.
 */
export class LivenessSession {
  private readonly options: SessionOptions

  private phase: LivenessState['phase'] = 'idle'
  private stepIndex = 0
  private framing: Framing = 'noFace'
  private reason: FailureReason | undefined

  private startedAt = 0
  private stepStartedAt = 0
  private lastT = 0
  private heldMs = 0
  private capturedThisRun = false
  private badFramingSince: number | null = null
  private badFramingKind: Framing | null = null
  /** The signal at the moment `center` completed — every later challenge is judged relative to it. */
  private baseline: FaceSignal | null = null
  /** Best value each step reached, in the challenge's own unit — the tuning telemetry. */
  private metrics: Record<string, StepMetric> = {}

  constructor(
    private readonly challenges: Challenge[],
    options: Partial<SessionOptions> = {},
    private readonly onEvent: (event: SessionEvent) => void = () => {},
  ) {
    if (challenges.length === 0) throw new Error('LivenessSession needs at least one challenge')
    this.options = { ...DEFAULT_SESSION_OPTIONS, ...options }
  }

  get state(): LivenessState {
    const challenge = this.challenges[this.stepIndex]
    return {
      phase: this.phase,
      stepIndex: this.stepIndex,
      stepCount: this.challenges.length,
      challenge: challenge ? challenge.name : null,
      holdProgress: this.currentHold > 0 ? Math.min(1, this.heldMs / this.currentHold) : 0,
      framing: this.framing,
      stepMetrics: { ...this.metrics },
      ...(this.reason ? { reason: this.reason } : {}),
    }
  }

  /** The neutral-frame signal, once `center` has completed. */
  get neutralBaseline(): FaceSignal | null {
    return this.baseline
  }

  start(now: number): LivenessState {
    this.phase = 'running'
    this.stepIndex = 0
    this.framing = 'noFace'
    this.reason = undefined
    this.startedAt = now
    this.stepStartedAt = now
    this.lastT = now
    this.resetHold()
    this.badFramingSince = null
    this.badFramingKind = null
    this.baseline = null
    this.metrics = {}
    return this.state
  }

  /** Hold required by the current step: the challenge's own, else the session's. */
  private get currentHold(): number {
    return this.challenges[this.stepIndex]?.holdMs ?? this.options.holdMs
  }

  /** Feed one camera frame. Returns the state the UI should render. */
  feed(signal: FaceSignal): LivenessState {
    if (this.phase !== 'running') return this.state

    const delta = Math.min(MAX_FRAME_DELTA_MS, Math.max(0, signal.t - this.lastT))
    this.lastT = signal.t

    if (signal.t - this.startedAt > this.options.totalTimeoutMs) return this.fail('timeout')
    if (signal.t - this.stepStartedAt > this.options.perStepTimeoutMs) return this.fail('timeout')

    this.framing = this.evaluateFraming(signal)

    if (this.framing === 'noFace' || this.framing === 'multipleFaces') {
      const expired = this.trackBadFraming(this.framing, signal.t)
      this.resetHold()
      if (expired) return this.fail(this.framing === 'multipleFaces' ? 'multipleFaces' : 'faceLost')
      return this.state
    }

    this.badFramingSince = null
    this.badFramingKind = null

    if (this.framing !== 'ok') {
      this.resetHold()
      return this.state
    }

    // Too soon after the step began to count — see `SessionOptions.minStepMs`.
    if (signal.t - this.stepStartedAt < this.options.minStepMs) return this.state

    const challenge = this.challenges[this.stepIndex]!
    this.recordMetric(challenge, signal)
    if (!challenge.isSatisfied(signal, this.baseline)) {
      this.resetHold()
      return this.state
    }

    this.heldMs += delta

    const hold = this.currentHold
    const progress = hold > 0 ? this.heldMs / hold : 1
    if (!this.capturedThisRun && progress >= this.options.captureAtProgress) {
      this.capturedThisRun = true
      this.onEvent({ type: 'capture', challenge: challenge.name, stepIndex: this.stepIndex })
    }

    if (this.heldMs >= hold) {
      // The centre step's last confirming frame becomes the baseline every
      // later challenge is measured against.
      if (this.stepIndex === 0) this.baseline = signal
      this.onEvent({ type: 'stepComplete', challenge: challenge.name, stepIndex: this.stepIndex })
      this.advance(signal.t)
    }

    return this.state
  }

  /** All steps captured; evidence is being uploaded. */
  notifyUploading(): LivenessState {
    if (this.phase === 'failed') return this.state
    this.phase = 'uploading'
    return this.state
  }

  /** The server has spoken. */
  notifyResult(passed: boolean, reason: FailureReason = 'cancelled'): LivenessState {
    this.phase = passed ? 'passed' : 'failed'
    this.reason = passed ? undefined : reason
    return this.state
  }

  abort(reason: FailureReason): LivenessState {
    if (this.phase === 'passed' || this.phase === 'failed') return this.state
    return this.fail(reason)
  }

  // -------------------------------------------------------------------------

  private advance(now: number): void {
    this.stepIndex += 1
    this.resetHold()
    this.stepStartedAt = now
    if (this.stepIndex >= this.challenges.length) {
      this.phase = 'uploading'
      this.onEvent({ type: 'complete' })
    }
  }

  private fail(reason: FailureReason): LivenessState {
    this.phase = 'failed'
    this.reason = reason
    this.resetHold()
    const challenge = this.challenges[this.stepIndex]
    this.onEvent({
      type: 'failed',
      reason,
      stepIndex: this.stepIndex,
      challenge: challenge ? challenge.name : null,
      stepMetrics: { ...this.metrics },
    })
    return this.state
  }

  /** Keep the best value this step has reached (max for 'above', min for 'below'). */
  private recordMetric(challenge: Challenge, signal: FaceSignal): void {
    const m = challenge.metric(signal, this.baseline)
    if (!m) return
    const key = `${this.stepIndex}:${challenge.name}`
    const prev = this.metrics[key]
    const better = !prev || (m.direction === 'above' ? m.value > prev.best : m.value < prev.best)
    if (better) {
      this.metrics[key] = { challenge: challenge.name, best: m.value, needed: m.needed, direction: m.direction, t: signal.t }
    } else if (prev) {
      // keep `needed` fresh (it can move once the baseline lands)
      prev.needed = m.needed
    }
  }

  private resetHold(): void {
    this.heldMs = 0
    this.capturedThisRun = false
  }

  /** Returns true once the bad framing has lasted longer than the grace period. */
  private trackBadFraming(kind: Framing, now: number): boolean {
    if (this.badFramingKind !== kind) {
      this.badFramingKind = kind
      this.badFramingSince = now
      return false
    }
    if (this.badFramingSince === null) {
      this.badFramingSince = now
      return false
    }
    return now - this.badFramingSince > this.options.faceLostGraceMs
  }

  /**
   * Position checks only — never pose. During a turn the head *should* be
   * rotated, so pose belongs to the challenge predicate, not here.
   */
  private evaluateFraming(signal: FaceSignal): Framing {
    if (signal.count === 0) return 'noFace'
    if (signal.count > 1) return 'multipleFaces'

    const { minFaceRatio, maxFaceRatio, maxOffCentre } = this.options
    if (signal.box.w < minFaceRatio) return 'tooFar'
    if (signal.box.w > maxFaceRatio) return 'tooClose'

    const cx = signal.box.x + signal.box.w / 2
    const cy = signal.box.y + signal.box.h / 2
    const offset = Math.hypot(cx - 0.5, cy - 0.5)
    if (offset > maxOffCentre) return 'offCentre'

    return 'ok'
  }
}

File: sdk/android/liveness/src/main/java/com/ekyc/liveness/engine/LivenessSession.kt
package com.ekyc.liveness.engine

import kotlin.math.abs
import kotlin.math.hypot
import kotlin.math.min

/** Largest time step we trust between two frames; guards against a stalled camera. */
private const val MAX_FRAME_DELTA_MS = 250L

/**
 * Drives the user through the challenge list and decides when the evidence
 * moment is. A faithful port of `LivenessSession` in `@ekyc/react-native-ekyc`
 * (136 tests on that side pin the mechanism; `LivenessSessionTest` here pins
 * the same cases).
 *
 * Pure Kotlin: no Android, no clocks. All time comes from `signal.t`.
 *
 * Each step is a *hold*: the challenge predicate must stay true for `holdMs`;
 * mid-hold a [SessionEvent.Capture] fires. Multi-phase challenges then need
 * their return movement (one confirming frame). Between challenges the head
 * must come back to the centre, so "left then right" is two movements through
 * the middle rather than one held swing.
 */
class LivenessSession(
    private val challenges: List<Challenge>,
    private val options: SessionOptions = SessionOptions(),
    private val onEvent: (SessionEvent) -> Unit = {},
) {
    init { require(challenges.isNotEmpty()) { "LivenessSession needs at least one challenge" } }

    private var phase = Phase.IDLE
    private var stepIndex = 0
    private var framing = Framing.NO_FACE
    private var reason: FailureReason? = null
    private var startedAt = 0L
    private var stepStartedAt = 0L
    private var lastT = 0L
    private var heldMs = 0L
    private var capturedThisRun = false
    private var badFramingSince: Long? = null
    private var badFramingKind: Framing? = null
    /** The signal at the moment `center` completed — every later challenge is judged relative to it. */
    var baseline: FaceSignal? = null
        private set
    private val metrics = LinkedHashMap<String, StepMetric>()
    private var stepPhase = 0
    private var memo = ChallengeMemo()
    private var awaitingRecenter = false

    val state: LivenessState
        get() {
            val c = challenges.getOrNull(stepIndex)
            val hold = currentHold
            return LivenessState(
                phase = phase,
                stepIndex = stepIndex,
                stepCount = challenges.size,
                challenge = c?.name,
                holdProgress = if (hold > 0) min(1f, heldMs.toFloat() / hold) else 0f,
                framing = framing,
                stepPhase = stepPhase,
                phaseCount = c?.phaseCount ?: 1,
                awaitingRecenter = awaitingRecenter,
                reason = reason,
                stepMetrics = LinkedHashMap(metrics),
            )
        }

    fun start(now: Long): LivenessState {
        phase = Phase.RUNNING
        stepIndex = 0
        framing = Framing.NO_FACE
        reason = null
        startedAt = now
        stepStartedAt = now
        lastT = now
        resetHold()
        badFramingSince = null
        badFramingKind = null
        baseline = null
        metrics.clear()
        stepPhase = 0
        memo = ChallengeMemo()
        awaitingRecenter = false
        return state
    }

    private val currentHold: Long
        get() = challenges.getOrNull(stepIndex)?.holdMs ?: options.holdMs

    /** Feed one detector frame. Returns the new state. */
    fun update(signal: FaceSignal): LivenessState {
        if (phase != Phase.RUNNING) return state
        val delta = (signal.t - lastT).coerceIn(0, MAX_FRAME_DELTA_MS)
        lastT = signal.t

        if (signal.t - startedAt > options.totalTimeoutMs) return fail(FailureReason.TIMEOUT)
        if (signal.t - stepStartedAt > options.perStepTimeoutMs) return fail(FailureReason.TIMEOUT)

        framing = evaluateFraming(signal)
        if (framing == Framing.NO_FACE || framing == Framing.MULTIPLE_FACES) {
            val expired = trackBadFraming(framing, signal.t)
            resetHold()
            if (expired) return fail(if (framing == Framing.MULTIPLE_FACES) FailureReason.MULTIPLE_FACES else FailureReason.FACE_LOST)
            return state
        }
        badFramingSince = null
        badFramingKind = null
        if (framing != Framing.OK) {
            resetHold()
            return state
        }
        if (signal.t - stepStartedAt < options.minStepMs) return state

        val challenge = challenges[stepIndex]
        if (awaitingRecenter) {
            if (!isRecentered(signal)) return state
            awaitingRecenter = false
        }

        recordMetric(challenge, signal)
        val satisfied = challenge.isSatisfied(signal, baseline, stepPhase, memo)

        if (stepPhase == challenge.capturePhase) {
            if (!satisfied) {
                resetHold()
                return state
            }
            heldMs += delta
            val hold = currentHold
            val progress = if (hold > 0) heldMs.toFloat() / hold else 1f
            if (!capturedThisRun && progress >= options.captureAtProgress) {
                capturedThisRun = true
                onEvent(SessionEvent.Capture(challenge.name, stepIndex))
            }
            if (heldMs < hold) return state
            if (stepIndex == 0) baseline = signal
            phaseDone(challenge, signal)
            return state
        }
        // Event phase (eyes open again, mouth closed, back to distance): one
        // confirming frame moves on; earlier phases are never undone.
        if (satisfied) phaseDone(challenge, signal)
        return state
    }

    fun abort(reason: FailureReason): LivenessState {
        if (phase == Phase.COMPLETED || phase == Phase.FAILED) return state
        return fail(reason)
    }

    private fun phaseDone(challenge: Challenge, signal: FaceSignal) {
        if (stepPhase < challenge.phaseCount - 1) {
            stepPhase += 1
            heldMs = 0
            return
        }
        onEvent(SessionEvent.StepComplete(challenge.name, stepIndex))
        advance(signal.t)
    }

    private fun isRecentered(signal: FaceSignal): Boolean {
        val b = baseline ?: return true
        val tol = options.recenterMaxDeg
        return abs(signal.yaw - b.yaw) <= tol && abs(signal.pitch - b.pitch) <= tol
    }

    private fun advance(now: Long) {
        stepIndex += 1
        resetHold()
        stepPhase = 0
        memo = ChallengeMemo()
        stepStartedAt = now
        val next = challenges.getOrNull(stepIndex)
        awaitingRecenter = options.requireRecenter && stepIndex >= 2 && next != null && next.requiresRecenter && baseline != null
        if (stepIndex >= challenges.size) {
            phase = Phase.COMPLETED
            onEvent(SessionEvent.Complete)
        }
    }

    private fun fail(reason: FailureReason): LivenessState {
        phase = Phase.FAILED
        this.reason = reason
        resetHold()
        onEvent(SessionEvent.Failed(reason, stepIndex, challenges.getOrNull(stepIndex)?.name))
        return state
    }

    private fun recordMetric(challenge: Challenge, signal: FaceSignal) {
        val m = challenge.metric(signal, baseline, stepPhase, memo)
        val key = if (stepPhase == 0) "$stepIndex:${challenge.name.wire}" else "$stepIndex:${challenge.name.wire}#$stepPhase"
        val prev = metrics[key]
        val better = prev == null || (if (m.direction == Direction.ABOVE) m.value > prev.best else m.value < prev.best)
        if (better) {
            metrics[key] = StepMetric(challenge.name, stepPhase, m.value, m.needed, m.direction, signal.t)
        } else if (prev != null && prev.needed != m.needed) {
            metrics[key] = prev.copy(needed = m.needed)
        }
    }

    private fun resetHold() {
        heldMs = 0
        capturedThisRun = false
    }

    private fun trackBadFraming(kind: Framing, now: Long): Boolean {
        if (badFramingKind != kind) {
            badFramingKind = kind
            badFramingSince = now
            return false
        }
        val since = badFramingSince ?: run { badFramingSince = now; return false }
        return now - since > options.faceLostGraceMs
    }

    /** Position checks only — never pose (during a turn the head *should* be rotated). */
    private fun evaluateFraming(s: FaceSignal): Framing {
        if (s.count == 0) return Framing.NO_FACE
        if (s.count > 1) return Framing.MULTIPLE_FACES
        if (s.box.w < options.minFaceRatio) return Framing.TOO_FAR
        if (s.box.w > options.maxFaceRatio) return Framing.TOO_CLOSE
        val cx = s.box.x + s.box.w / 2
        val cy = s.box.y + s.box.h / 2
        if (hypot(cx - 0.5f, cy - 0.5f) > options.maxOffCentre) return Framing.OFF_CENTRE
        return Framing.OK
    }
}
